lib/src/lifecycle_module.dart

```diff
@@ -487,83 +488,79 @@ abstract class LifecycleModule extends SimpleModule with Disposable {
   /// throw a [StateError].
   @protected
-  Future<Null> loadChildModule(LifecycleModule childModule) {
+  Future<Null> loadChildModule(LifecycleModule childModule) async {
     if (isOrWillBeDisposed) {
       return _buildDisposedOrDisposingResponse(methodName: 'loadChildModule');
     }
 
     if (_childModules.contains(childModule)) {
-      return Future.value(null);
+      return null;
     }
 
     if (isUnloaded || isUnloading) {
       var stateLabel = isUnloaded ? 'unloaded' : 'unloading';
-      return Future.error(
-          StateError('Cannot load child module when module is $stateLabel'));
+      throw StateError('Cannot load child module when module is $stateLabel');
     }
 
-    final completer = Completer<Null>();
-    onWillLoadChildModule(childModule).then((LifecycleModule _) async {
-      _willLoadChildModuleController.add(childModule);
-
-      final childModuleWillUnloadSub = listenToStream(
-          childModule.willUnload, _onChildModuleWillUnload,
-          onError: _willUnloadChildModuleController.addError);
-      final childModuleDidUnloadSub = listenToStream(
-          childModule.didUnload, _onChildModuleDidUnload,
-          onError: (error, stackTrace) =>
-              _didUnloadChildModuleController.addError);
-
-      // The child module may not reach an unloaded state successfully, but
-      // should always eventually be disposed. For this reason, we listen for
-      // its disposal before removing it from the list of child modules.
-      // ignore: unawaited_futures
-      childModule.didDispose.then((_) {
-        _childModules.remove(childModule);
-      });
-
-      try {
-        manageDisposable(childModule);
-        _childModules.add(childModule);
-        childModule._parentContext = _loadContext;
-
-        await childModule.load();
-        try {
-          await onDidLoadChildModule(childModule);
-        } catch (error, stackTrace) {
-          _logger.severe(
-            'Exception in onDidLoadChildModule ($name)',
-            error,
-            stackTrace,
-          );
-          rethrow;
-        }
-        _didLoadChildModuleController.add(childModule);
-        completer.complete();
-      } catch (error, stackTrace) {
-        // If the child module failed to load, we can dispose of it and cleanup
-        // any state/subscriptions related to it.
-        _childModules.remove(childModule);
-        await childModule.dispose();
-        await childModuleWillUnloadSub.cancel();
-        await childModuleDidUnloadSub.cancel();
-
-        _didLoadChildModuleController.addError(error, stackTrace);
-        completer.completeError(error, stackTrace);
-      } finally {
-        childModule._parentContext = null;
-      }
-    }).catchError((Object error, StackTrace stackTrace) {
+    try {
+      await onWillLoadChildModule(childModule);
+    } catch (error, stackTrace) {
       _logger.severe(
         'Exception in onWillLoadChildModule ($name)',
```

Comment:
Wow having this error handling right next to the actual code that triggered the error is so much easier to follow